### ISSUE TITLE
Behat failures don't fail the build

### ISSUE THREAD

**Issue by timhunt** [moodlehq/moodle-plugin-ci#164]
This is very odd, and more than a little annoying:

https://github.com/moodleou/moodle-quiz_gradingstudents/runs/5512440974?check_suite_focus=true#step:17:134

```
Running single behat site:
Moodle 3.9.12+ (Build: 20220311), 7c52eaaf180ba34534c3b4106fe18ac7a02f0f20
Php: 7.2.34.28.20.04.1.1, mariadb: 10.7.3, OS: Linux 5.11.0-1028-azure x86_64
Server OS "Linux", Browser: "chrome"
Started at 11-03-2022, 22:48
P----------------P----------------------------------------P----------- 70
--------P--------------------P------------------

--- Pending steps:

001 Scenario: report with no attempts                    # /home/runner/work/moodle-quiz_gradingstudents/moodle-quiz_gradingstudents/moodle/mod/quiz/report/gradingstudents/tests/behat/gradingstudents.feature:57
      Given the following "custom profile fields" exist: # behat_data_generators::the_following_entities_exist()
        "custom profile fields" is not a known type of entity that can be generated.
```

It could not generate "custom profile fields" in this version of Moodle, but rather than getting reported as a failure, it got reported as 'pending'. And, that did not fail the build.

(So, then, we released an incompatible version.)

So, as a start, I am reporting the bug here.

**Comment by stronk7** [comment]
Hi Tim,

just was looking to the [behat documentation](https://behat.org/en/latest/user_guide/context/definitions.html#pending-steps) and it seems to be by design... from the information in the link:

> When a definition method throws any Exception (except PendingException) during execution, the step is marked as failed. 
> ...
> If you use --strict option with Behat, pending steps will cause Behat to exit with 1 code.

And from the behat CLI help:

> -strict                           Passes only if all tests are explicitly passing.

So, maybe the only option here... is to add support for that `--strict` option so you can add it to your runs?

I'm not sure we should add it by default (it's quite possible to have skipped tests, for example), and using that option (not tested!) seems to lead to exit with error code also for them.

**Comment by timhunt** [comment]
Thanks for finding that.

Right, I think defaulting --strict to on for Moodle plugin ci runs is right. If you have pending things, then your code is not fully working, so not ready for a release.

**Comment by stronk7** [comment]
>  I think defaulting --strict to on for Moodle plugin ci runs is right

Oh, sorry. I meant the opposite, we cannot enable it by default because it also fails with skipped steps. And skipped steps is something acceptable (don't run this if that... sort of things).

So, the only possibility I see is that we add support for the behat `--strict` option and, whoever wants to use it... there it is.

Ciao :-)

**Comment by timhunt** [comment]
Oh! that sucks.

Still having pending steps in release code should be a blocker to the release, and so should fail the build.

However, having skipped steps, because some tests are not relevant in some contexts is OK, I think.

So, really, it seems like the only way to get what we want would be one of:
* Send some more specific options upstreap to Behat (--fail-on-pending, --fail-on-skipped) and use them.
* Hack the moodle-plugin-ci wrapper to parse the console output, rather than just basing the CI result on the behat cli result.

**Comment by scara** [comment]
Hi @timhunt,
> Send some more specific options upstreap to Behat (--fail-on-pending, --fail-on-skipped) and use them.

FYI https://github.com/moodlehq/moodle-plugin-ci/pull/165.
Not sure about your `--fail-on-pending`: typo for `warning`?

HTH,
Matteo

**Comment by timhunt** [comment]
No. 'Pending steps' are a specific thing in Behat (https://behat.org/en/latest/user_guide/context/definitions.html#pending-steps) and we want an option which specifically causes the build to fail if any are found. The name is right.

**Comment by jrchamp** [comment]
If `--strict` truly causes problems for **skipped** tests (which I can't find in the documentation), then perhaps something similar to [MDL-57669](https://tracker.moodle.org/browse/MDL-57669) could be done when support for erroring on undefined steps was added.

**Comment by scara** [comment]
> If `--strict` truly causes problems for **skipped** tests (which I can't find in the documentation), then perhaps something similar to [MDL-57669](https://tracker.moodle.org/browse/MDL-57669) could be done when support for erroring on undefined steps was added.

For (my) convenience: https://github.com/moodle/moodle/commit/ada21cb0c836110f335ce3e3034038881091e985

HTH,
Matteo

**Comment by andrewnicols** [comment]
Skipped tests are a Moodle thing. We have a need for them, as do others,
and so we are following the guidance on how to skip them without causing
test failures.

We may be able to modify that to not cause fails at all… I think I even
experimented with it.

https://github.com/moodle/moodle/blob/998c632b5da9366681ff55fd53ab23b63ccc01b7/lib/behat/extension/Moodle/BehatExtension/EventDispatcher/Tester/ChainedStepTester.php#L126

**Comment by stronk7** [comment]
> If --strict truly causes problems for skipped tests (which I can't find in the documentation)

They do, I tried [some comments above](https://github.com/moodlehq/moodle-plugin-ci/issues/164#issuecomment-1108895938) without (passing) and with (failing) `--strict` . They lead to strictly fail anything not being a passed step.

So,

1. Or we are able to make `--strict` not to be so strict (and allow skipped steps to pass).
2. Or we apply some change making the Pending steps to lead to failure. Ideally without hacks looking for output, but properly within the extension.

(I personally feel that 2 is the correct solution, though haven't looked if it's achievable)

Ciao :-)
